### DISCUSSION TITLE
Complejidad: ecuacion de recurrencia

### DISCUSSION THREAD

**Discussion by clangoi** [IIC2133-PUC/2024-1#62] [Preguntas materia] [answered]
Hola, estaba viendo la clase de propiedades de QuickSort, y cuando se define la complejidad de tiempo ponen la ecuacion de recurrencia, pero no estoy tan segura de como calcular dicha ecuación.
![image](https://github.com/IIC2133-PUC/2024-1/assets/53914794/37f81318-db5b-4efd-b572-39ff31f0507a)

**Comment by CarloGauss33** [comment, marked as answer]
Hola @clangoi.

### Peor caso
(recordando que es cuando se seleccionan particiones de 1 elemento): Se demuestra por induccion constructiva de forma simple

Dejo el desarrollo

la idea es simplemente expandir la recursión hasta identificar el patron de que se asemeja a `\sum_{i=0}^{N-1} (N-i)` y por ende aplicando la formula de gauss para calcular 
![image](https://github.com/IIC2133-PUC/2024-1/assets/45949389/e92f18d3-74bc-4355-bc0b-b72df56757eb)

### Mejor caso: 

**Por Teo Maestro**
Dado que se cumple que el tamaño de las particiones decae geometricamente (y que el número de subproblemas es constante), se puede aplicar teorema maestro (como se vio en Discretas).

_(Teo maestro para recordar)_
<img width="627" alt="image" src="https://github.com/IIC2133-PUC/2024-1/assets/45949389/d62b82a4-8084-4710-9eb3-98f4d229a165">

En este caso `f(n) = n`, `a, b = 2`
![image](https://github.com/IIC2133-PUC/2024-1/assets/45949389/a740cd95-af88-4753-89e7-94c7afc73a3a)
Y por ende cae en el segundo caso de O(nLogn)

**Por inducción**
Por otra parte tambien se puede por inducción. Voy a tratar de escribirlo según discussions me deja:
Tomando la ecuación y desarrollando

`T(n) = n + 2T(n/2) = n + 2*(  n/2 + 2T(n/4) )`
`= n + 2*(n/2 + 2*(n/4 + T(n/8)....`

Lo que se puede reducir  a la expresión
(Considerando que logN - 1 es la cantidad maxima de veces que N se puede dividir)
![image](https://github.com/IIC2133-PUC/2024-1/assets/45949389/5f67b691-bbe0-40b3-ae1d-04f055310f13)
